File: eNote/eNote.Application/PagedResult.cs
namespace eNote.Application
{
    public class PagedResult<T>
    {
        public IReadOnlyList<T> Items { get; init; } = Array.Empty<T>();

        public int Page { get; init; }
        public int PageSize { get; init; }

        public int ReturnedCount => Items.Count;
        public int? TotalCount { get; init; }

        public int? TotalPages => TotalCount.HasValue ? (int)Math.Ceiling((double)TotalCount / PageSize) : null;
    }
}

File: UI/enote_core/lib/paging/paged_result.dart
class PagedResult<T> {
  final List<T> items;
  final int page;
  final int pageSize;
  final int? totalCount;

  PagedResult({
    required this.items,
    this.page = 1,
    this.pageSize = 20,
    this.totalCount,
  });

  int? get totalPages =>
      totalCount != null && pageSize > 0
          ? (totalCount! / pageSize).ceil()
          : null;
  bool get hasNext =>
      totalCount != null ? (page * pageSize) < totalCount! : items.length == pageSize;
  bool get hasPrevious => page > 1;

}
